frontend/catalyst/compilation_pipelines.py

```diff
@@ -550,9 +546,102 @@ def __call__(self, *args, **kwargs):
         else:
             args = CompiledFunction.promote_arguments(self.c_sig, r_sig, *args)
 
+        if any(isinstance(arg, jax.core.Tracer) for arg in args):
+            # Only compile a derivative version of the compiled function when needed.
+            if self.jaxed_qfunc is None:
+                self.jaxed_qfunc = JAX_QJIT(self)
+
+            return self.jaxed_qfunc(*args, **kwargs)
+
         return self.compiled_function(*args, **kwargs)
 
 
+class JAX_QJIT:
+    """Wrapper class around :class:`~.QJIT` that enables compatibility with JAX transformations.
+
+    The primary mechanism through which this is effected is by wrapping the invocation of the QJIT
+    object inside a JAX ``pure_callback``. Additionally, a custom JVP is defined in order to support
+    JAX-based differentiation, which is itself a ``pure_callback`` around a second QJIT object which
+    invokes :func:`~.grad` on the original function. Using this class thus incurs additional
+    compilation time.
+
+    Args:
+        qfunc (QJIT): the compiled quantum function object to wrap
+    """
+
+    def __init__(self, qfunc):
+        @jax.custom_jvp
+        def jaxed_qfunc(*args, **kwargs):
+            results = self.wrap_callback(qfunc, *args, **kwargs)
+            if len(results) == 1:
+                results = results[0]
+            return results
+
+        self.qfunc = qfunc
+        self.jaxed_qfunc = jaxed_qfunc
+        self._deriv_qfunc = None  # compiled on demand
+        jaxed_qfunc.defjvp(self.compute_jvp)
+
+    @property
+    def deriv_qfunc(self):
+        """Compile a function computing the derivate of the wrapped QJIT function."""
+        if self._deriv_qfunc is not None:
+            return self._deriv_qfunc
+
+        def deriv_wrapper(*args, **kwargs):
+            # Only consider differentiable arguments (i.e. of type floating point).
+            all_diff_args = [i for i in range(len(args)) if args[i].dtype.kind == "f"]
+
+            return catalyst.grad(self.qfunc, argnum=all_diff_args)(*args, **kwargs)
+
+        annotations = {}
+        signature = inspect.signature(self.qfunc)
+        for idx, (arg_name, param) in enumerate(signature.parameters.items()):
+            param._annotation = self.qfunc.c_sig[idx]
```

Comment:
Weird that pylint didn't warn about this when to code was located in the `__init__` function 🤔

Reply:
https://github.com/PennyLaneAI/catalyst/pull/96/commits/c1f649449c63472dc3deb449f7e5a7abcf386c38